Author: Niko Tyni <ntyni@debian.org>
Subject: FTBS: tests fail triggered by libwww-perl 5.813 to 5.820
Forwarded: no
Last-Update: 2011-03-22
Bug-Debian: http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=509457
Reviewed-by: Nicholas Bamber <nicholas@periapt.co.uk>
--- a/Apache-Test/lib/Apache/TestRequest.pm
+++ b/Apache-Test/lib/Apache/TestRequest.pm
@@ -261,6 +261,11 @@
     return $self->get_basic_credentials(@_);
 }
 
+sub credentials {
+    my $self = shift;
+    return $self->get_basic_credentials(@_);
+}
+
 sub get_basic_credentials {
     my($self, $realm, $uri, $proxy) = @_;
 
