Description: add a conditional define for Apache 2.4 (-D APACHE24)
Origin: CPAN RT
Bug: https://rt.cpan.org/Public/Bug/Display.html?id=87620
Author: Michael Schout
Reviewed-by: Ivan Kohler <ivan@freeside.biz>
Last-Update: 2014-06-27

--- a/Apache-Test/lib/Apache/TestServer.pm
+++ b/Apache-Test/lib/Apache/TestServer.pm
@@ -85,6 +85,15 @@ sub post_config {
         $self->{rev} = 0; # unknown
     }
 
+    ($self->{revminor}) = $self->{version} =~ m|/\d\.(\d)|;
+
+    if ($self->{revminor}) {
+        debug "Matched Apache revminor $self->{version} $self->{revminor}";
+    }
+    else {
+        $self->{revminor} = 0;
+    }
+
     $self;
 }
 
@@ -126,7 +135,14 @@ sub pid_file {
 
 sub dversion {
     my $self = shift;
-    "-D APACHE$self->{rev}";
+
+    my $dv = "-D APACHE$self->{rev}";
+
+    if ($self->{rev} == 2 and $self->{revminor} >= 4) {
+        $dv .= " -D APACHE24";
+    }
+
+    return $dv;
 }
 
 sub config_defines {
